File: examples/matlab/svar/wegner2025-section-5-1.m
clear; 
clc; 

addpath("~/Documents/repos/tca-matlab-toolbox/");
addpath("~/Documents/repos/tca-matlab-toolbox/plotting");
addpath("~/Documents/repos/tca-matlab-toolbox/models");

data = readtable("./data.csv");

%-------------------------------------------------------------------------------
% GERTLER KARADI
%-------------------------------------------------------------------------------

keepCols = {'mp1_tc', 'ffr', 'ygap', 'infl', 'lpcom'};
dataGK = data(:, keepCols);

% VAR using 4 lags, a constant, and a linear trend
model = VAR(dataGK, 4, 'trendExponents', 0:1);
model.fit();
% Internal instrument 
% choosing ffr to be the normalising variable
method = InternalInstrument('ffr');
% computing IRFs
irfObj = model.IRF(40, 'identificationMethod', method);
irfs = irfObj.getIrfArray();
irfs = irfs(:, 1, :) * 0.25;
% Defining transmission matrix
transmissionOrder = {'mp1_tc', 'ffr', 'ygap', 'infl', 'lpcom'};
% Defining the transmission channels
channelNonContemp = model.notThrough('ffr', 0, transmissionOrder);
channelContemp = model.through('ffr', 0', transmissionOrder);
% Transmission effects
effectsNonContemp = model.transmission(1, channelNonContemp, transmissionOrder, 40, 'identificationMethod', method);
effectsContemp = model.transmission(1, channelContemp, transmissionOrder, 40, 'identificationMethod', method);
% adjusting for shock size
effectsNonContemp = effectsNonContemp * 0.25;
effectsContemp = effectsContemp * 0.25;
% The effects should perfectly decompose the total effects
max(vec(irfs - effectsContemp - effectsNonContemp))
% Plotting the decomposition
channelNames = ["Contemporaneous", "Non-Contemporaneous"];
cellChannelEffects = {effectsContemp, effectsNonContemp};
% for ffr
fig = plotDecomposition(2, irfs, cellChannelEffects, channelNames);
set(fig, 'Units', 'inches', 'Position', [1 1 10 5]);
exportgraphics(fig, 'gk-ffr.png', 'Resolution', 300);
% for inflation
fig = plotDecomposition(4, irfs, cellChannelEffects, channelNames);
set(fig, 'Units', 'inches', 'Position', [1 1 10 5]);
exportgraphics(fig, 'gk-infl.png', 'Resolution', 300);

%-------------------------------------------------------------------------------
% ROMER AND ROMER
%-------------------------------------------------------------------------------

keepCols = {'rr_3', 'ffr', 'ygap', 'infl', 'lpcom'};
dataRR = data(:, keepCols);

% VAR using 4 lags, a constant, and a linear trend
model = VAR(dataRR, 4, 'trendExponents', 0:1);
model.fit();
% Internal instrument 
% choosing ffr to be the normalising variable
method = InternalInstrument('ffr');
% computing IRFs
irfObj = model.IRF(40, 'identificationMethod', method);
irfs = irfObj.getIrfArray();
irfs = irfs(:, 1, :) * 0.25;
% Defining transmission matrix
transmissionOrder = {'rr_3', 'ffr', 'ygap', 'infl', 'lpcom'};
% Defining the transmission channels
channelNonContemp = model.notThrough('ffr', 0, transmissionOrder);
channelContemp = model.through('ffr', 0', transmissionOrder);
% Transmission effects
effectsNonContemp = model.transmission(1, channelNonContemp, transmissionOrder, 40, 'identificationMethod', method);
effectsContemp = model.transmission(1, channelContemp, transmissionOrder, 40, 'identificationMethod', method);
% adjusting for shock size
effectsNonContemp = effectsNonContemp * 0.25;
effectsContemp = effectsContemp * 0.25;
% The effects should perfectly decompose the total effects
max(vec(irfs - effectsContemp - effectsNonContemp))
% Plotting the decomposition
channelNames = ["Contemporaneous", "Non-Contemporaneous"];
cellChannelEffects = {effectsContemp, effectsNonContemp};
% for ffr
fig = plotDecomposition(2, irfs, cellChannelEffects, channelNames);
set(fig, 'Units', 'inches', 'Position', [1 1 10 5]);
exportgraphics(fig, 'rr-ffr.png', 'Resolution', 300);
% for inflation
fig = plotDecomposition(4, irfs, cellChannelEffects, channelNames);
set(fig, 'Units', 'inches', 'Position', [1 1 10 5]);
exportgraphics(fig, 'rr-infl.png', 'Resolution', 300);

% Minor differences between the decompositions here and the ones in the paper 
% are due to how we obtain the orthogonal IRFs. In the paper we estimate 
% orthogonal IRFs using a VAR excluding the internal instrument. Here we 
% include the internal instrument. At this point we have no preference for one 
% method over the other. The differences are minor anyways. 

File: examples/julia/svar/wegner2025-section-5-1.jl
using TransmissionChannelAnalysis
using DataFrames, CSV
using CairoMakie, Makie  # for plotting

data = DataFrame(CSV.File("./data.csv"))

#-------------------------------------------------------------------------------
# GERTLER KARADI
#-------------------------------------------------------------------------------

# Selecting the required variables
data_gk = select(data, :mp1_tc, :ffr, :ygap, :infl, :lpcom)

# Defining the VAR model with a constant and linear trend
model = VAR(data_gk, 4; trend_exponents=0:1)
fit!(model)

# We are using an internal instrument (the first variable) and normalise it 
# by the contemporaneous response of `ffr`.
method = InternalInstrument(:ffr)
# Computing IRFs using the internal instrument. Normalising the response to a 
# 25bps increase in ffr. 
irf_obj = IRF(model, method, 40)
irfs = irf_obj.irfs[:, 1:1, :] * 0.25

# Defining the transmission matrix / order
transmission_order = [:mp1_tc, :ffr, :ygap, :infl, :lpcom]

# Defining the transmission channels
channel_noncontemp = not_through(model, :ffr, 0:0, transmission_order)
channel_contemp = !channel_noncontemp

# Computing transmission effects
effects_noncontemp = transmission(model, method, 1, channel_noncontemp, transmission_order, 40)
effects_contemp = transmission(model, method, 1, channel_contemp, transmission_order, 40)
# Adjusting for the shock size
effects_noncontemp .*= 0.25
effects_contemp .*= 0.25
# The effects should perfectly decompose the total effects
isapprox(irfs, effects_noncontemp + effects_contemp; atol=sqrt(eps()))

# Plotting only one outcome
teffects = [effects_contemp, effects_noncontemp]
channel_names = ["Contemporaneous", "Non-contemporaneous"]
plot_decomposition(2, irfs[:, 1:1, :], teffects, channel_names; legend=true)

# Plotting the decomposition for all outcomes in one figure
fig = Figure(;size=(1000, 300));
ax = Axis(fig[1, 1]; title="FFR")
plot_decomposition!(ax, 2, irfs[:, 1:1, :], teffects);
ax2 = Axis(fig[1, 2]; title="Inflation");
plot_decomposition!(ax2, 4, irfs[:, 1:1, :], teffects);
add_decomposition_legend!(fig[2, :], channel_names)
fig

save("./gk_both.png", fig)

#-------------------------------------------------------------------------------
# ROMER AND ROMER
#-------------------------------------------------------------------------------

# Selecting the required variables
data_rr = select(data, :rr_3, :ffr, :ygap, :infl, :lpcom)

# Defining the VAR model with a constant and linear trend
model = VAR(data_rr, 4; trend_exponents=0:1)
fit!(model)

# We are using an internal instrument (the first variable) and normalise it 
# by the contemporaneous response of `ffr`.
method = InternalInstrument(:ffr)
# Computing IRFs using the internal instrument. Normalising the response to a 
# 25bps increase in ffr. 
irf_obj = IRF(model, method, 40)
irfs = irf_obj.irfs[:, 1:1, :] * 0.25

# Defining the transmission matrix / order
transmission_order = [:rr_3, :ffr, :ygap, :infl, :lpcom]

# Defining the transmission channels
channel_noncontemp = not_through(model, :ffr, 0:0, transmission_order)
channel_contemp = through(model, :ffr, 0:0, transmission_order)

# Computing transmission effects
effects_noncontemp = transmission(model, method, 1, channel_noncontemp, transmission_order, 40)
effects_contemp = transmission(model, method, 1, channel_contemp, transmission_order, 40)
# Adjusting for the shock size
effects_noncontemp .*= 0.25
effects_contemp .*= 0.25
# The effects should perfectly decompose the total effects
isapprox(irfs, effects_noncontemp + effects_contemp; atol=sqrt(eps()))

# Plotting the decomposition
teffects = [effects_contemp, effects_noncontemp]
channel_names = ["Contemporaneous", "Non-contemporaneous"]
fig = Figure(;size=(1000, 300));
ax = Axis(fig[1, 1]; title="FFR")
plot_decomposition!(ax, 2, irfs[:, 1:1, :], teffects);
ax2 = Axis(fig[1, 2]; title="Inflation");
plot_decomposition!(ax2, 4, irfs[:, 1:1, :], teffects);
add_decomposition_legend!(fig[2, :], channel_names)
fig

save("./rr_both.png", fig)
